walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,175 @@
+        }
+        else if(mandatory_attendees.size()== 0){
+            return optional_times;
+        }
+        if(optional_attendees.size() != 0 && mandatory_attendees.size()!= 0 && common_times.size() == 0){
+            System.out.println("mandatory times returned");
+            return mandatory_times;
+        }
+        else{
+            System.out.println("common_times returned");
+            return common_times;
+        }
+    
+    }
+  else {
+      return mandatory_times;
+      }
   }
+
+
+   public Collection<TimeRange> findMandatoryAttendeeTimes(Collection<Event> events, MeetingRequest request){
+       Collection<TimeRange> mandatory_times = new ArrayList<TimeRange>();
+       Collection<String> mandatory_attendees = request.getAttendees();
+       mandatory_times.add(TimeRange.WHOLE_DAY);
+      
+       for (Event event : events){
+        boolean mandatory_attendee_overlap = false;
+        Set<String> event_attendees = event.getAttendees();
+        for (String attendee : event_attendees){
+          if (mandatory_attendees.contains(attendee)){
+            mandatory_attendee_overlap = true;
+            break;
+            }
+        }
+        if (mandatory_attendee_overlap ==  true){ 
+            ArrayList<TimeRange> to_remove = new ArrayList<TimeRange>();
+            ArrayList<TimeRange> to_add= new ArrayList<TimeRange>();
+            for(TimeRange freeslot : mandatory_times){
+                if (freeslot.overlaps(event.getWhen())){
+                    if (event.getWhen().start()<=freeslot.end() && event.getWhen().start()>=freeslot.start()){
+                        if (freeslot.start() != event.getWhen().start()){
+                          to_add.add(TimeRange.fromStartEnd(freeslot.start(), event.getWhen().start(),false));
+                        }
+                    }
+                    if (event.getWhen().end()<= freeslot.end() && event.getWhen().end()>= freeslot.start()){
+                        if(event.getWhen().end() != freeslot.end()){
+                          to_add.add(TimeRange.fromStartEnd(event.getWhen().end(), freeslot.end(),false));
+                        }
+                    }
+                    to_remove.add(freeslot);
+                }
+            }
+            mandatory_times.removeAll(to_remove);
+            mandatory_times.addAll(to_add);
+        }
+    }
+    ArrayList<TimeRange> mandatory_duration_too_short = new ArrayList<TimeRange>();
+    for (TimeRange freeslot : mandatory_times){
+        if (freeslot.duration()< request.getDuration()){
+            mandatory_duration_too_short.add(freeslot);
+        }
+    }
+    mandatory_times.removeAll(mandatory_duration_too_short);
+    return mandatory_times;
+   }
+
+
+   public Collection<TimeRange> findOptionalAttendeeTimes(Collection<Event> events, MeetingRequest request){
+    Collection<TimeRange> optional_times = new ArrayList<TimeRange>();
+    Collection<String> optional_attendees = request.getOptionalAttendees();
```

Comment:
is there any difference between findOptionalAttendeeTimes and findMandatoryAttendeeTimes besides whether the attendee list is retrieved from request.getOptionalAttendees vs request.getAttendees? If not, can you think of a way to combine these two methods?

Reply:
yep! i made it into one function